fms/models/qwen.py

```diff
@@ -0,0 +1,782 @@
+            max_seq_len_cached (dict[Optional[torch.device], int]): _description_
+        """
+        # remove meta tensors from cached_freqs
+        for dev in list(cached_freqs.keys()):
+            for alp in list(cached_freqs[dev].keys()):
+                if cached_freqs[dev][alp].device == torch.device("meta"):
+                    del cached_freqs[dev][alp]
+                    if len(cached_freqs[dev]) == 0:
+                        del cached_freqs[dev]
+                        del max_seq_len_cached[dev]
+
+    def post_init(self):
+        """_summary_
+        """
+        # This function is called in `get_model` after the model is
+        # fully initalized on the correct device
+
+        self._clean_up_rot_emb_cache(
+            self.rot_emb.cached_freqs,
+            self.rot_emb.max_seq_len_cached,
+        )
+
+        # init RoPE on the right device(s)
+        for device in set(
+            [param.device for param in self.parameters()]
+            + [buffer.device for buffer in self.buffers()]
+        ):
+            self.rot_emb.compute_freqs_cis(device, self.config.max_expected_seq_len)
+
+    def forward(
+        self,
+        x_in,
+        mask=None,
+        position_ids=None,
+        past_key_value_states=None,
+        use_cache=False,
+        attn_algorithm=None,
+    ):
+        """_summary_
+
+        Args:
+            x_in (_type_): _description_
+            mask (_type_, optional): _description_. Defaults to None.
+            position_ids (_type_, optional): _description_. Defaults to None.
+            past_key_value_states (_type_, optional): _description_. Defaults to None.
+            use_cache (bool, optional): _description_. Defaults to False.
+            attn_algorithm (_type_, optional): _description_. Defaults to None.
+
+        Returns:
+            _type_: _description_
+        """
+        # Embed the given vocabulary indices using the given attention mask, with pre-/post-norm
+        # and dropout as specified
+        # x_in: batch_size x seq_len
+        # mask: batch_size x seq_len x seq_len
+        # bias: nheads x seq_len x seq_len
+        if past_key_value_states is None or len(past_key_value_states) == 0:
+            past_key_value_states = [None for _ in range(len(self.layers))]
+
+        qlen = x_in.size(1)
+        klen = x_in.size(1)
+
+        # if we are using the cache, the key length needs to be extended with the past keys length
+        if use_cache and past_key_value_states[0] is not None:
+            klen += past_key_value_states[0][0].size(-2)
+
+        # if mask is none, we need to specify causal mask
+        if mask is None:
+            # we are caching and can assume all 1s in the mask
+            if use_cache and klen != 1 and qlen == 1:
+                # b x h x qlen x kvlen
+                is_causal_mask = False
+            else:
+                is_causal_mask = True
+        else:
+            is_causal_mask = False
```

Comment:
I think all this logic can be cleaned up, look at granite or llama in the latest fms